### PR TITLE
Extension Simplification

### LINKED ISSUE #1523: Disable existing plugins through labextension
Is it possible to disable certain existing plugins of JLab through extensions? Eg: If I wish to disable the launcher plugin on the left panel, so that I only see files,commands,running etc.

### LINKED ISSUE #2011: Plan for Extension Handling
After several attempts, we have finally come to the conclusion that it is not feasible to handle extensions without using node to build the assets on the server.  Our latest attempt, #1973, was to use a WebPack extension to build individual bundles, and then assemble the bundles on the server, using semver to deduplicate packages.  What we soon realized is it is impossible to know for sure which modules from an upstream npm package will be consumed, so we cannot effectively deduplicate packages at the package level without having the entire package.

The proposed solution is to bundle the entire `node_modules` tree with an extension, and then merge the trees and build the final WebPack bundle on the server.  We rely on `npm` to deduplicate the packages in their entirety, and WebPack to bundle the assets.  

Note that the above are only required to add new extensions to the system.  Bundles can be distributed that contain a preconfigured set of extensions, and the user can be given the opportunity to enable or disable individual extensions in the bundle (cf #1523).

The requirement to have `node` installed on the user's machine can be alleviated by using the `node` package from `conda`, or we could use a standalone node executable builder such as [nexe](https://github.com/nexe/nexe) that could be distributed as a Python wheel.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #1523.  Fixes #2011.

Notes on the differences:

From the user perspective:
- no more `--py`, `user/sys` or `enable/disable` of extensions
- can now manually set `page_config` in `etc/labconfig/pageconfig.json`, including `ignorePlugins` which is an array of token names to ignore.

From a core developer perspective:
- Run `jupyter lab --dev-mode` to ensure using a local build vs `/share`
- Run `npm run build:main` or `npm watch:main` while developing
- Can build `all-packages` to populate all of the packages simultaneously (default for `npm run build`).

From an extension author perspective:
- The extension is an npm package (anything that you can `npm install`)- it must only depend on public packages unless it bundles them
- It must have `jupyterlab: { extension: True }` metadata.
- `jupyter labextension link` installs the extension as a link (without using `npm link`)
- It must only include ES5 CommonJS, CSS, and some other formats that we support.

### CODE REVIEW

**Comment by blink1073** [comment]
Progress:

```
ssilvester: ~/anaconda/share/jupyter/lab
$ ls build/
0662c22d4e544cecf2bc66ee4c761ed6.svg   b06871f281fee6b241d60582ae9369b9.ttf
26a0370f727f4667dfbf94b8ec576dc4.svg   cdbbf3faea3d6d5b7c7c8a27736254df.gif
28f2f5e6ac027acb7fbd41f7e02ebf90.gif   fcad2f97acb3c868bd1ca7629d36d80e.svg
2b609d364e2e1d67ba33a81c58d9eb8a.svg   fee66e712a8a08eef5805a46892932ad.woff
5e9fca8da62c2b939b204d8883012664.gif   main.bundle.js
674f50d287a8c48dc19ba404d20fe713.eot   main.bundle.js.map
912ec66d7572ff821749319396470bde.svg   main.css
93441bee5fcf0a095b35aa85f722b06c.svg   main.css.map
af7ae505a9eed503f8b8e6982036873e.woff2
```

**Comment by blink1073** [comment]
The above required manually installing `@jupyterlab/default` theme in the build directory.  The next step is to add the default extensions to the `package.json` template.

**Comment by blink1073** [comment]
Sweet, easier than expected.

**Comment by blink1073** [comment]
Next steps are to validate this with `jupyterlab_geojson`, and then rectify the "fallback" local build.

**Comment by blink1073** [comment]
Success!  cf https://github.com/jupyterlab/jupyterlab_geojson/pull/33

<image src="https://cloud.githubusercontent.com/assets/2096628/24933096/2ccb22cc-1eda-11e7-9516-b7d17ff81eea.png" width=400>

**Comment by blink1073** [comment]
File structure:

```bash
ssilvester: ~/anaconda/share/jupyter/lab
$ ls
build             index.js          package.json
cache             node_modules      webpack.config.js

ssilvester: ~/anaconda/share/jupyter/lab
$ ls cache
jupyterlab_geojson_labextension-0.17.0.tgz
```

**Comment by blink1073** [comment]
The files used in `<repo>/jupyterlab` now form the basis of `share/jupyterlab`, and they use identical workflows.  The difference is that the local build will always use the local `packages/` and a fixed set of extensions.

**Comment by blink1073** [comment]
Given the above and the `npm link` proof of concept in https://github.com/jupyterlab/jupyterlab_geojson/pull/33#issuecomment-293540632, I'd say the core functionality is done.  Next steps are to create the entry scripts, update our build scripts, and then handle config.

**Comment by blink1073** [comment]
(and update tests).

**Comment by blink1073** [comment]
I need to write up some docs and do a pass through, but this is done.

**Comment by blink1073** [comment]
Switching back to the previous entry point style next:

```
jupyter lab
jupyter labextension install
jupyter labextension uninstall
jupyter labextension list
jupyter labbuild
```

Also looking into updating to automatically run `labbuild` before launching when we detect that the `package.json` hash differs from what we have stored from the last run.

**Comment by blink1073** [comment]
We can load a placeholder page that auto-refreshes on a timer that is served until the build is complete.

**Comment by sccolbert** [comment]
That could be a good enough first-pass. We can probably do better in the future loading a page with some spinner graphics and a bit of JS which polls the server for completion, adds the real app script tag, then transitions to the app.

**Comment by blink1073** [comment]
Actually it is probably easier to do that the first time, using a simple CSS load indicator https://www.w3schools.com/howto/howto_css_loader.asp and a separate tornado handler for the build status.

**Comment by blink1073** [comment]
For now, we always build when installing or uninstalling an extension, we can handle deferred builds later.

**Comment by blink1073** [comment]
I was about to write up the docs on how we don't yet handle running against master but it makes more sense to do that part now since it informs the other workflows.  Thoughts:

- `/share/lab` is always linked back to `packages/` for the core extensions - so there is no confusion about which we are using

- `jupyter labextension link` (defaults to cwd unless a path is given)
    - `npm install --save` the package directly to get it rectified
    - create the symlink back to the package
    - add the package to the `symlinks` metadata
    - create the symlinks to `packages/` in its `node_modules`

- `jupyter lab build` 
  - ensure there is a `node_modules` (run `npm install` if not)
  - ensure that we have all of the symlinks
  -  run webpack

**Comment by blink1073** [comment]
@gnestor ^^

**Comment by blink1073** [comment]
And `jupyter labextension unlink` would remove the package from `package.json` and remove the symlinks in the package's folder.

**Comment by jasongrout** [comment]
I think you could also make a case that `jupyter labextension build` is an appropriate name - you're building that extension into your current lab. Or perhaps `jupyter labextension buildlab` or something.

**Comment by blink1073** [comment]
Build can be run at any time (as can the new `jupyter lab clean`), so I think it makes sense to have those separate from `labextensions`.

**Comment by blink1073** [comment]
Left to do: handle sym-linking of core extensions, and tests (unit and against `jupyterlab_geojson`).

**Comment by gnestor** [comment]
@blink1073 Ok, is this different than https://github.com/jupyterlab/jupyterlab_geojson/pull/33#issuecomment-293540632. It would link jupyterlab's `packages` in the extension's `node_modules`? Vs link the extension in the `packages` directory? Either way, I really like this workflow. I know that there can be issues with symlinks, but as long as we are using `npm link` it should be fine.

**Comment by blink1073** [comment]
I had started down the path of manually sym-linking the paths to avoid using the global `npm link` location, but that path seems to brittle.  I think the right approach is to use `npm link` but use our own `npm link` location (using the `prefix` npm config).

**Comment by gnestor** [comment]
That sounds like a good solution. The `npm link` location is the global `node_modules` path, right? Is there any issue with using that?

**Comment by blink1073** [comment]
We can't go global because jlab is per user environment (conda or virtualenv).

**Comment by blink1073** [comment]
Some notes:
 
- `npm pack` includes all of the nested dependencies in `bundledDependencies` - this cannot include local file paths (private packages).
- `npm install` works with no internet connection when installing a tarball or local directory.
- `npm pack` does not work on a lerna package with `bundleDependencies` due to the non-standard handling of the symlinks (the dependencies are not included).

- Using the above knowledge, we can provide a script to extension authors that
  - verifies the `jupyterlab: extension: true` metadata
  - ensures all of the `dependencies` are in `bundleDependencies`
  - ensures that all of them are available to pack using [read-installed](https://github.com/npm/read-installed) from npm.

This would allow `conda` and `wheel` packages to include all of the data needed to install.
From a technical standpoint we can support both bundled and standard packages on the user's machine: they would be need to have an internet connection once they use one or more unbundled packages.

The question remains whether we should enforce that all valid extensions should also have valid `bundleDependencies` so they can be consumed in an offline environment.

**Comment by blink1073** [comment]
https://www.npmjs.com/package/npm-bundle could be used as the bundling tool.

**Comment by blink1073** [comment]
More notes:

- Config should be entirely outside of package.json
- page_config.json - ignoredPlugins, etc.
- build_config.json - build location, linked extensions,
installed extensions

- jupyterlab/node_modules is used as the basis when cleaning
- we then pull in package.json and install based on config
data
- we use `npm link` for the core packages and any 
explicitly symlinked packages using a custom prefix


Bundling options:
- not on the roadmap for npm core: https://github.com/npm/npm/issues/4210#issuecomment-221707747
- npm-bundle uses standard npm commands has not been maintained in over a year and
does not support scoped packages
- npmbox is actively updated and has an open PR for scoped packages, but uses non-standard installation `npmunbox foo.npmbox`.  

My thought is that `jupyter labextension install foo` should handle `.npmbox` files in addition to the standard `npm install`-able targets.   I do not think we should enforce any specific tool (or bundling at all), but we can have a standard `deploy` script that uses `npmbox`.

**Comment by blink1073** [comment]
Let's not use a custom prefix for linking, it will only cause confusion.  Additionally, `npm link --prefix` doesn't work but `export NPM_CONFIG_PREFIX=foo; npm link` does work, so it is brittle.

**Comment by blink1073** [comment]
Refocusing, because can't satisfy all of the use cases using the same approach.  

Case 1: User installing an extension
- Runs `jupyter labextension install foo`which installs the extension in `<sys-prefix>/share/jupyter/lab` and builds the application.   

Case 2: Admin creating a shared build
- Installs the appropriate extensions.  Users get a `build_config.json` that points to the shared build.

Case 3: Extension author
- Checks out the desired version of JupyterLab.  Installs compatible npm packages in their repository (this can be scripted).  Runs `jupyter labextension install foo` (optionally with --link).  Link is accomplished by re-installing the extension prior to build, not by using `npm link`.  We provide pre-releases of the JupyterLab subcomponents to ease this workflow.

Case 4: Extension author that needs to run against master (for concurrent changes)
- Adds the package to `lerna.json` in the JupyterLab repo and adds it as a core extension.

Case 5: Core developer working on JupyterLab itself.
- Runs `jupyter lab --dev` to launch from the `jupyterlab/` built sources (the ones that are shipped).

**Comment by blink1073** [comment]
@jasongrout, as of e21abda I'd say the code changes are done, working on docs now.

**Comment by blink1073** [comment]
cc @willingc, major changes to extension docs in d80f857 (formatting fixes in b3b080a).

File: tutorial/extensions_dev.md [hunk 1]
```diff
@@ -66,3 +66,52 @@ throughout.  **Phosphor messages** are a *many-to-one* interaction that allows
 information like resize events to flow through the widget hierarchy in
 the application.  **Phosphor signals** are a *one-to-many* interaction that allow
 listeners to react to changes in an observed object.
+
+
+## Extension Authoring
+An Extension is a valid npm package that meets the following criteria:
```
Comment (ian-r-rose) [inline tutorial/extensions_dev.md]: Perhaps gloss or link to a definition of npm for people more accustomed to the Python ecosystem?

File: tutorial/extensions_dev.md [hunk 2]
```diff
@@ -66,3 +66,52 @@ throughout.  **Phosphor messages** are a *many-to-one* interaction that allows
...
+The package should export EMCAScript 5 compatible JavaScript.  It can 
+import CSS using the syntax `require('foo.css')`.  The CSS files
+can also import CSS from other packages using the syntax 
+`@import url('~foo/index.css')`, where `foo` is the name of the package.
+
+The following file types are also supported (both in JavaScript and CSS):
+json, html, jpg, png, gif, svg, js.map, woff2, ttf, eot.
+
+If your package uses any other file type it must be converted to one of 
+the above types.  If you JavaScript is written in any other dialect than
```
Comment (ian-r-rose) [inline tutorial/extensions_dev.md]: you -> your

File: tutorial/extensions_dev.md [hunk 3]
```diff
@@ -66,3 +66,52 @@ throughout.  **Phosphor messages** are a *many-to-one* interaction that allows
...
+EMCAScript 5 it must be converted using an appropriate tool.
+
+The package must only rely on other packages that are bundled with your extension or available on npm.org.
+
+If you publish your extension on npm.org, users will be able to 
+install it as simply `jupyter labextension install <foo>`, where
+`<foo>` is the name of the published npm package.  You can alternatively
+provide a script that runs `jupyter labextension install` against a 
+local folder path on the user's machine or a provided tarball.  Any
+valid `npm install` specifier can be used in `jupyter labextension install`.
```
Comment (ian-r-rose) [inline tutorial/extensions_dev.md]: By "specifier" do you mean command line arguments?

File: tutorial/extensions_user.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+The currently installed extensions can be listed by running the command:
+
+```
+jupyter labextension list
+```
+
+An installed extension can be uninstalled by running the command:
+
+```
+jupyter labextension <bar>
```
Comment (ian-r-rose) [inline tutorial/extensions_user.md]: This looks like it is missing an action.

File: tutorial/extensions_dev.md [hunk 5]
```diff
@@ -66,3 +66,52 @@ throughout.  **Phosphor messages** are a *many-to-one* interaction that allows
...
+`@import url('~foo/index.css')`, where `foo` is the name of the package.
+
+The following file types are also supported (both in JavaScript and CSS):
+json, html, jpg, png, gif, svg, js.map, woff2, ttf, eot.
+
+If your package uses any other file type it must be converted to one of 
+the above types.  If you JavaScript is written in any other dialect than
+EMCAScript 5 it must be converted using an appropriate tool.
+
+The package must only rely on other packages that are bundled with your extension or available on npm.org.
```
Comment (jasongrout) [inline tutorial/extensions_dev.md]: I found that (for development purposes), if my package depended on an absolute path on my computer, it was fine. I would probably just delete this line - it's kind of a given that it should be npm-installable.

File: tutorial/extensions_dev.md (hunk 3 above)
Comment (jasongrout) [inline tutorial/extensions_dev.md]: I don't think so. I think it means that you could do something like `jupyter labextension install foo@3.0.0`, `jupyter labextension install myfile.tar.gz`, etc. Basically, if you could do `npm install X`, you can do `jupyter labextension install X`.

Comment (blink1073) [inline tutorial/extensions_dev.md]: What Jason said (updated).

**Comment by blink1073** [comment]
Comments addressed, WIP status removed.

**Comment by blink1073** [comment]
Actually, I need to update `contributing.md` as well.

**Comment by jasongrout** [comment]
> This is awesome. Thanks!

Yes! Thanks!

For the record, it worked for upgrading ipywidgets with pretty minimal changes: https://github.com/jupyter-widgets/ipywidgets/pull/1303

**Comment by SylvainCorlay** [comment]
@jasongrout @blink1073 it seems that installing labextensions only works with dev installations of jupyterlab, since installation from  pip / conda packages don't include the package.json file.

**Comment by jasongrout** [comment]
I just released 0.20.3 which should fix that problem.

**Comment by hainm** [comment]
@blink1073 is it true that user need to have `npm` to install labextension?

**Comment by blink1073** [comment]
Yes npm (or perhaps yarn) will be required to install extensions.

**Pull request merged**